src/SharpLearning.Backend.TensorFlow.Test/TensorFlowRawMnistDeepTest.cs

```diff
@@ -119,32 +148,87 @@ public void MnistDeep()
+                    var testBatchSize = 100;
+                    Debug.Assert(data.TestTargets.Data.Length % testBatchSize == 0); // Or we need to do something that handles remaining samples
+                    var rawTestBatchEnumerator = data.CreateTestBatchEnumerator(testBatchSize);
+                    var testBatchEnumerator = Convert(rawTestBatchEnumerator, ClassCount);
+                    var totalCount = 0;
+                    var correctCount = 0;
+                    while (testBatchEnumerator.MoveNext())
+                    {
+                        var (testInputBatch, testLabelBatch) = testBatchEnumerator.CurrentBatch();
+
+                        TFTensor testInputBatchTensor = TFTensor.FromBuffer(new TFShape(testBatchSize, FeatureCount), testInputBatch, 0, testInputBatch.Length);
+                        TFTensor testLabelBatchTensor = TFTensor.FromBuffer(new TFShape(testBatchSize, ClassCount), testLabelBatch, 0, testLabelBatch.Length);
+
+                        TFTensor evaluatedCorrectPrediction = session.GetRunner()
+                            .AddInput(x, testInputBatchTensor)
+                            .AddInput(y_, testLabelBatchTensor)
+                            .AddInput(keep_prob, test_dropout_keep_prob)
+                            .Run(correctPrediction);
 
-                    Log($"Accuracy {acc}");
-                    Assert.AreEqual(0.2029999941587448, acc);
+                        var ps = (bool[])evaluatedCorrectPrediction.GetValue();
+
+                        totalCount += ps.Length;
+                        foreach (var p in ps)
+                        {
+                            if (p) { ++correctCount; }
+                        }
+                    }
+
+                    var incorrectCount = totalCount - correctCount;
+                    var acc = correctCount / (float)totalCount;
+#endif
+                    var testTime_ms = s.Elapsed.TotalMilliseconds;
+                    s.Restart();
+
+                    Log($"Accuracy {acc} Initialize {initializeTime_ms,6:F1} Train {trainTime_ms,6:F1} Test {testTime_ms,6:F1} ");
+                    Assert.AreEqual(0.0979999974370003, acc, 0.00000001); // This is what C# currently computes
+                    Assert.AreEqual(0.2461999952793121, acc); // This is what equivalent python computes
                 }
             }
         }
 
+        private static IFlatBatchFeaturesTargetEnumerator<float, float> Convert(
+            IFlatBatchFeaturesTargetEnumerator<byte, byte> rawTrainBatchEnumerator, int classCount)
+        {
+            var trainBatchEnumerator = rawTrainBatchEnumerator
+                // TODO: We really want to do this before batch enumerator so same as for tests..
+                //.From().To(fb => (float)fb, tb => (int)tb);
+                //.Feature().To(fb => (float)fb)
+                .Feature().To(fb => fb * (1.0f / byte.MaxValue))
+                .Target().ToOneHot(t => t, 1.0f, classCount);
```

Comment:
@mdabros this fluent pattern for specifying conversions on data is a bit problematic and the whole concept of BatchEnumerator does not scale well to cases with more than one input or outputs etc. It is, however, strongly typed. NOTE: I still do the divide by 255 on the input and not in the graph yet, since that is not what the TF python examples do.

Reply:
Its perfectly fine to do the normalization when reading the data. 
When it is "out in the open" like this, it is clear to the user that the reader normalizes the data. Then one can simply decide if normalization should be part of the graph/model or if it should be done as a separate pre-step to the model.
I have not looked at the code for the BatchEnumerator in details yet, but the example of usage looks good. In time, we should of course be able to support cases with more inputs and outputs.